### PR TITLE
feat: rules and skills

### PR DESCRIPTION
**Pull request by marissahuysentruyt**
## Description

This PR adds Cursor agent skills and refactors existing AI instructions into the rule/skill structure within the `.cursor` directory. It introduces reusable, documented skills for common workflows (accessibility, docs, commits, migration, handoff, testing) and an always-applied deep-understanding rule, and updates .cursor documentation accordingly.

**NOTE:** not all skills have been tested. We can continue to refine these documents, however, after we've tried to use them in realistic workflows and settings.

**New skills**
- [`accessibility-compliance`](https://skills.sh/wshobson/agents/accessibility-compliance): WCAG 2.2–oriented guidance with ARIA patterns, mobile accessibility, and reference docs (aria-patterns.md, mobile-accessibility.md, wcag-guidelines.md).
- [`ask-questions`](https://skills.sh/trailofbits/skills/ask-questions-if-underspecified): When and how to clarify requirements before implementing.
- `component-migration-rendering-styling`: Creating rendering-and-styling migration analysis for 2nd-gen components.
- `contributor-docs-nav`: Running the CONTRIBUTOR-DOCS nav script (breadcrumbs, TOCs, link checks).
- [`conventional-commits`](https://skills.sh/marcelorodrigo/agent-skills/conventional-commit): Writing conventional commit messages.
- `deep-understanding`: Scoping, deep read, persistent research.md, pause for review, then implement (used by the always-applied rule).
- `documentation`: Following Adobe content writing standards when writing docs.
- [`explain-code`](https://code.claude.com/docs/en/skills#getting-started): Explaining code with diagrams and analogies. 
- [`session-handoff`](https://skills.sh/softaworks/agent-toolkit/session-handoff): Creating handoff docs, resume checklists, and scripts (create_handoff.py, validate_handoff.py, list_handoffs.py, check_staleness.py) plus evals and references.
- [`test-driven-development`](https://skills.sh/obra/superpowers/test-driven-development): TDD workflow and testing anti-patterns reference.

☝️ Any skills linked above were externally written/created, but may have been tweaked to fit our needs more specifically (i.e. writing to a `.cursor` folder instead of `.claude`) 

**Rules and docs**
- Deep-understanding rule: Always applied; requires research and a written artifact before planning or code.
- AI agent instructions moved into the rule/skill model.
- `.cursor/README.md`: Updated to describe newer rules and the new skills (with brief descriptions and when to use each).

## Motivation and context

- Consistency: Centralize AI behavior in skills and rules so Cursor follows the same patterns (docs, a11y, commits, migration, handoffs, TDD).
- Discoverability: Document skills in the README so contributors know what’s available and when to use it.
- Quality: Deep-understanding as a rule reduces “signature-level” work and encourages research and a reviewable artifact before implementation.
- Maintainability: Session-handoff skill and scripts improve context transfer and long-running work across sessions.

## Open questions

- Currently, our AI rules assume that everyone is using Cursor. This might mean that the rules and skills won't automatically be used if a contributor/team mate _isn't_ using Cursor. 

❓ Should we also create a small section in the `.cursor/README` or even the project's README to explain that non-Cursor users can copy the `rules` and `skills` contents into their own local IDE-specific directory (like `.vscode`) so they get the rules/skills automatically as well?

## Related issue(s)

- addresses SWC-1503 (CSS conventions)
- addresses SWC-1505 (accessibility requirements)
- partially addresses SWC-1498 and SWC-1506 (washing machine workflow), as well as SWC-1537 (AI-friendly documentation epic)
- addresses SWC-1605

## Author's checklist

- [x] I have read the CONTRIBUTING and PULL_REQUESTS documents.
- [ ] I have reviewed the Accessibility Practices for this feature, see: Aria Practices
- [ ] I have added automated tests to cover my changes.
- [ ] I have included a well-written changeset if my change needs to be published.
- [x] I have included updated documentation if my change required it.

## Reviewer's checklist

- [x] Includes a Github Issue with appropriate flag or Jira ticket number without a link
- [ ] Includes thoughtfully written changeset if changes suggested include patch, minor, or major features

## Manual review test cases

- [ ] Skills and rules appear in repo
    - Open the repo in Cursor with this branch.
    - Confirm .cursor/skills/* and .cursor/rules/* are present
    - Expect no broken paths or missing files
    - Reading the rules `.mdc` files and skills `SKILL.md` files still make understandable sense to humans. (i.e. does the skill make enough sense that you understand what the agent should do?)

- [ ] Documentation is coherent
    - Read .cursor/README.md sections for rules and skills.
    - Expect clear, consistent descriptions and correct references to skill paths.

### Additional validation

Feel free to test any of the new skills or rules! For instance...

- [ ] Session-handoff skill runs & works
    - With the branch pulled down locally, open a past chat (preferably one where you ran the context window quite high).
    - If possible in that chat, @ mention the `session-handoff` skill or say something to the affect of "context is getting full" or "create handoff." You should see the agent working with something like this:

Screenshot 2026-02-24 at 2 15 35 PM

  - Verify a session handoff is created (named something like "2026-02-24-141544-cursor-rules-skills-docs.md") is created in a `.cursor/handoffs` directory. Once the agent is done, it should give you a brief summary of the handoff doc and steps for the next session/chat.
  - Start a new session or chat, and resume your work by saying something like "resume from handoff" or "load handoff." You should see something like this once the agent has gotten the context from the old session:

Screenshot 2026-02-24 at 2 22 52 PM

### CODE REVIEW

File: .cursor/skills/contributor-docs-nav/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+### Maintainer: update the script
+
+- **When**: Naming/convention changes, breadcrumb/TOC format changes, new edge cases, or display-name rule changes.
+- **Process**: Review change → update `update-nav.js` → test on full tree → update docs and report.
+
+## Full instructions
+
+For complete Operator and Maintainer workflows, debugging, link-verification handling, script behavior details, and the testing checklist, read:
+
+**CONTRIBUTOR-DOCS/01_contributor-guides/07_authoring-contributor-docs/01_ai-agent-instructions.md**
```
Comment (rise-erpelding) [inline .cursor/skills/contributor-docs-nav/SKILL.md]: This doesn't have to be a decision we make here and now, but bringing up here the thought that I had before about where this `01_ai-agent-instructions.md` should live. To me, this file is a skill that we had before we knew to call it a skill, and we could move its contents up into this file to keep things a bit clearer.

File: .cursor/skills/component-migration-rendering-styling/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+- Create all files on the **original spectrum-web-components branch** where the session started
+- Verify git branch before analyzing any file; perform line-by-line comparisons (don’t assume templates are identical)
+- Ask clarifying questions for uncertain mappings instead of guessing
+
+## Full instructions
+
+For the exact document structure, heading levels, collapsible sections, CSS⇒SWC mapping table format (columns, status values, ordering), DOM comparison format, and output format notes, read:
+
+**CONTRIBUTOR-DOCS/03_project-planning/02_workstreams/02_2nd-gen-component-migration/02_step-by-step/01_analyze-rendering-and-styling/cursor_prompt.md**
```
Comment (rise-erpelding) [inline .cursor/skills/component-migration-rendering-styling/SKILL.md]: Same question for this as well as with `01_ai-agent-instructions`, is there an appetite for moving the reusable prompt up and officially calling it a skill?

File: .cursor/skills/session-handoff/evals/model-expectations.md
Comment (rise-erpelding) [inline .cursor/skills/session-handoff/evals/model-expectations.md:1]: These model explanations are really interesting, did you generate these or did they come from somewhere else?

File: .cursor/skills/session-handoff/evals/setup_test_env.py
Comment (rise-erpelding) [inline .cursor/skills/session-handoff/evals/setup_test_env.py:1]: Makes sense to me to put scripts here, but now I have a follow up question to my earlier question about moving the CONTRIBUTOR-DOCS agent instructions into skills, if we did that, should we move the update-nav.js script (and the other script there) into `.cursor/` as well?

File: .cursor/skills/session-handoff/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+### Step 1: Generate Scaffold
+
+Run the smart scaffold script to create a pre-filled handoff document:
+
+```bash
+python scripts/create_handoff.py [task-slug]
+```
+
+Example: `python scripts/create_handoff.py implementing-user-auth`
```
Comment (rise-erpelding) [inline .cursor/skills/session-handoff/SKILL.md]: Should these examples specify the full path? For instance

```suggestion
Example: `python .cursor/skills/session-handoff/scripts/create_handoff.py implementing-user-auth`
```

When I had agents (via Codex & Cursor) see if there was anything here that was problematic, they both flagged this. My understanding is that the skills would be applied from the repo root, not from where the skill is located, thus why the agents are flagging this as a potential issue.

I assumed that session handoff would still work fine because AI could figure this out - which is true. I tried a session handoff (which it kinda bungled because I was on your branch and not the branch I had done the rest of the stuff from the chat I asked it to create session-handoff from), and then I asked it some questions about how it did it, and it confirmed that it had to search for the scripts then run them from the full path from `.cursor/`.

FWIW it's also pointing out the Python version, it's running `python3` - that feels less significant.

File: .cursor/skills/session-handoff/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+**For continuation handoffs** (linking to previous work):
+
+```bash
+python scripts/create_handoff.py "auth-part-2" --continues-from 2024-01-15-auth.md
+```
+
+The script will:
+
+- Create `.claude/handoffs/` directory if needed
```
Comment (rise-erpelding) [inline .cursor/skills/session-handoff/SKILL.md]: Is there a reason this is `.claude`?

But this reminds me that you also raise an excellent point in the PR description about "what if someone's not using Cursor"

Also, would we/should we add the handoffs directory to `.gitignore` or just trust people/agents to not be committing that in PRs? I think I lean toward it being a worthwhile add to the `.gitignore`.

File: .cursor/skills/session-handoff/scripts/validate_handoff.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    todos = re.findall(r'\[TODO:[^\]]*\]', content)
+    return len(todos) == 0, todos
+
+
+def check_required_sections(content: str) -> tuple[bool, list[str]]:
+    """Check that required sections exist and have content."""
+    missing = []
+    for section in REQUIRED_SECTIONS:
+        # Look for section header
+        pattern = rf'(?:^|\n)##?\s*{re.escape(section)}'
```
Comment (rise-erpelding) [inline .cursor/skills/session-handoff/scripts/validate_handoff.py]: This will match at any heading level rather than just # or ##

```suggestion
        # Look for section header at any level (#, ##, ###, etc.) to match template
        pattern = rf'(?:^|\n)#+\s*{re.escape(section)}'
```

File: .cursor/skills/session-handoff/scripts/validate_handoff.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        match = re.search(pattern, content, re.IGNORECASE)
+        if not match:
+            missing.append(f"{section} (missing)")
+        else:
+            # Check if section has meaningful content (not just placeholder)
+            section_start = match.end()
+            next_section = re.search(r'\n##?\s+', content[section_start:])
```
Comment (rise-erpelding) [inline .cursor/skills/session-handoff/scripts/validate_handoff.py]: Also here. You could ask Cursor to double-check the scripts on your end before modifying to test if we definitely need these changes.

```suggestion
            # Next section is any markdown header (one or more #) so content boundary is correct
            next_section = re.search(r'\n#+\s+', content[section_start:])
```

File: .cursor/skills/conventional-commits/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+The header is required. Scope is optional. All lines must stay under 100 characters.
+
+## Commit Types
+
+| Type       | Purpose                               |
+| ---------- | ------------------------------------- |
+| `build`    | Build system or CI changes            |
+| `chore`    | Routine maintenance tasks             |
+| `ci`       | Continuous integration configuration  |
+| `deps`     | Dependency updates                    |
```
Comment (rise-erpelding) [inline .cursor/skills/conventional-commits/SKILL.md]: Do we ever use this `deps` type? (I feel like our team tends to use `chore` instead.)

We don't list `deps` in [`branch-naming.mdc`](https://github.com/adobe/spectrum-web-components/blob/marissahuysentruyt/feat-agent-skills/.cursor/rules/branch-naming.mdc#L24-L34) or in [`.cursor/config.json`](https://github.com/adobe/spectrum-web-components/blob/marissahuysentruyt/feat-agent-skills/.cursor/config.json#L15-L26).

I'm wondering if there should be a single source of truth for all the types as well? Like this skill and the branch-naming rule should both refer to the config.json instead of re-listing all the types again?

File: .cursor/skills/session-handoff/evals/model-expectations.md
Comment (marissahuysentruyt) [reply .cursor/skills/session-handoff/evals/model-expectations.md:1]: These were directly from the session handoff install: https://github.com/softaworks/agent-toolkit/blob/main/skills/session-handoff/evals/model-expectations.md

and they're probably already out of date 😭

File: .cursor/skills/session-handoff/SKILL.md (hunk 4 above)
Comment (marissahuysentruyt) [reply .cursor/skills/session-handoff/SKILL.md]: whoops nope! I thought I changed that to `.cursor` so that was just a miss here.

I like the idea of adding it to the gitignore! 💡

File: .cursor/rules/deep-understanding.mdc [hunk 8]
```diff
@@ -0,0 +1,16 @@
+---
+description: Before writing any code, do deep research on the relevant part of the codebase and write findings to a persistent file (e.g. research.md). Applies at the start of every session and whenever tackling a new area.
+alwaysApply: true
```
Comment (blunteshwar) [inline .cursor/rules/deep-understanding.mdc]: Perhaps we could use `apply Intelligently` instead? Sometimes, users only need a simple request, like creating a regex pattern. If we always apply this rule, we might be wasting the tokens and poisoning the context unnecessarily.

File: .cursor/skills/accessibility-compliance/references/aria-patterns.md [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+        </div>
+    );
+}
+```
+
+### Tabs
+
+```tsx
+function Tabs({ tabs }) {
+    const [activeIndex, setActiveIndex] = useState(0);
```
Comment (blunteshwar) [inline .cursor/skills/accessibility-compliance/references/aria-patterns.md]: How about we tweak the examples to better fit the LIT context since we’re using LIT in SWC? Right now, many examples seem to be using React.

File: .cursor/skills/session-handoff/scripts/check_staleness.py [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+Analyzes:
+- Time since handoff was created
+- Git commits since handoff
+- Files that changed since handoff
+- Branch divergence
+- Modified files status
+
+Usage:
+    python check_staleness.py <handoff-file>
+    python check_staleness.py .claude/handoffs/2024-01-15-143022-auth.md
```
Comment (blunteshwar) [inline .cursor/skills/session-handoff/scripts/check_staleness.py]: Perhaps we should use .cursor/handoffs/**

File: .cursor/skills/session-handoff/scripts/create_handoff.py [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+    )
+    if success and staged:
+        info["staged_files"] = staged.split("\n")
+
+    return info
+
+
+def find_previous_handoffs(project_path: str) -> list[dict]:
+    """Find existing handoffs in the project."""
+    handoffs_dir = Path(project_path) / ".claude" / "handoffs"
```
Comment (blunteshwar) [inline .cursor/skills/session-handoff/scripts/create_handoff.py]: Instead of `.cursor` `.claude` is being used for handoffs path. Perhaps we should use `.cursor/handoffs/*` here and everywhere.

File: .cursor/README.md [hunk 12]
```diff
@@ -1,8 +1,8 @@
-# Cursor rules documentation
+# Cursor documentation
 
-This directory contains rules that Cursor uses to enforce consistent formatting and structure in our codebase.
+This directory contains rules and skills that Cursor uses to enforce consistent formatting and structure in our codebase.
 
-## Rule structure
+## Rules
 
 Rules defined in the `config.json` follow this structure:
```
Comment (blunteshwar) [inline .cursor/README.md:7]: Can you confirm if there exists a file named `config.json`?

Comment (marissahuysentruyt) [reply .cursor/README.md:7]: Yep! https://github.com/adobe/spectrum-web-components/blob/main/.cursor/config.json

File: .cursor/skills/session-handoff/SKILL.md (hunk 3 above)
Comment (marissahuysentruyt) [reply .cursor/skills/session-handoff/SKILL.md]: Those should be fixed in several places in the file now!

File: .cursor/skills/contributor-docs-nav/SKILL.md (hunk 1 above)
Comment (marissahuysentruyt) [reply .cursor/skills/contributor-docs-nav/SKILL.md]: Agreed! I've moved it up to this directory under `reference/ai-agent-instructions.md`. 

When I was chatting with Cursor it brought up a good point that the original directory in CONTRIBUTOR-DOCS was a little confusing/deceiving since there were both computer/AI docs alongside human docs. I think moving it makes sense.

File: .cursor/skills/session-handoff/evals/setup_test_env.py
Comment (marissahuysentruyt) [reply .cursor/skills/session-handoff/evals/setup_test_env.py:1]: I did this! We have a `scripts` and `reference` directory in `contributor-docs-nav` now. 

Open for discussion the naming. 🤔

File: .cursor/skills/component-migration-rendering-styling/SKILL.md (hunk 2 above)
Comment (marissahuysentruyt) [reply .cursor/skills/component-migration-rendering-styling/SKILL.md]: yyyyessss doing it.

File: .cursor/skills/accessibility-compliance/references/aria-patterns.md (hunk 9 above)
Comment (marissahuysentruyt) [reply .cursor/skills/accessibility-compliance/references/aria-patterns.md]: Let me know what you think of the new examples! Cursor rewrote them.

File: .cursor/skills/conventional-commits/SKILL.md (hunk 7 above)
Comment (marissahuysentruyt) [reply .cursor/skills/conventional-commits/SKILL.md]: I removed the deps, and added some more documentation around using the config.json.

File: .cursor/skills/accessibility-compliance/references/mobile-accessibility.md [hunk 13]
```diff
@@ -0,0 +1,541 @@
...
+            aria-label={label}
+            className="relative p-3" // Creates 44x44 touch area
+        >
+            <span className="block w-5 h-5">{icon}</span>
+        </button>
+    );
+}
+```
+
+## iOS VoiceOver
```
Comment (cdransf) [inline .cursor/skills/accessibility-compliance/references/mobile-accessibility.md]: Should we remove the iOS/Android/native section of this skill?

**Comment by blunteshwar** [comment]
Once ci is green I think we are good to merge
